cmd/snap-update-ns/bootstrap.c

```diff
@@ -247,6 +247,91 @@ int validate_snap_name(const char* snap_name)
     return 0;
 }
 
+static int instance_key_validate(const char *instance_key)
+{
+	// NOTE: see snap.ValidateInstanceName for reference of a valid instance key
+	// format
+
+	// Ensure that name is not NULL
+	if (instance_key == NULL) {
+		bootstrap_msg = "instance key cannot be NULL";
+		return -1;
+	}
+
+	const char *p = instance_key;
+	int n = 0;
+	while (*p != '\0') {
+		int m = 0;
+		m = skip_lowercase_letters(&p);
+		if (m > 0) {
+			n += m;
+			continue;
+		}
+		m = skip_digits(&p);
+		if (m > 0) {
+			n += m;
+			continue;
+		}
+		bootstrap_msg = "instance key must use lower case letters or digits";
+		return -1;
+	}
+
+	if (n == 0) {
+		bootstrap_msg = "instance key must contain at least one letter or digit";
+		return -1;
+	} else if (n > 10) {
+		bootstrap_msg = "instance key must be shorter than 10 characters";
+		return -1;
+	}
+
+	return 0;
+}
+
+// validate_instance_name performs full validation of the given snap instance name.
+int validate_instance_name(const char* instance_name)
+{
+    // NOTE: This function should be synchronized with the two other
+    // implementations: sc_instance_name_validate and snap.ValidateInstanceName.
+
+    if (instance_name == NULL) {
+        bootstrap_msg = "snap instance name cannot be NULL";
+        return -1;
+    }
+
```

Comment:
My take on this using strsep (which I find slightly easier to read):
```
    // 40 char snap_name + 10 char instance_key + 1 NULL + 1 extra to detect
    // overflows
    char s[52] = {0,};
    strncpy(s, instance_name, sizeof(s));

    char *t[3] = {s};
    const char *snap_name = strsep(t, "_");
    const char *instance_key = strsep(t, "_");
    const char *third_separator = strsep(t, "_");
    if (third_separator != NULL) {
          return -1;
    }
    if (validate_snap_name(snap_name) < 0) {
       return -1;
    }
    if (instance_key != NULL && instance_key_validate(instance_key) < 0) {
       return -1;
    return 0;
```

Reply:
I do like the simplicity of mvo's implementation here, but I might change it slightly to initialize t as `char *t[1] = {s};`

Reply:
Updated, thanks for the patch and suggestions.